spacy_llm/tasks/ner.py

```diff
@@ -18,41 +20,38 @@ def find_substrings(
     text: str,
     substrings: Iterable[str],
     *,
+    lang: str = "xx",
     case_sensitive: bool = False,
     single_match: bool = False,
 ) -> Iterable[Tuple[int, int]]:
     """Given a list of substrings, find their character start and end positions
     in a text"""
-
-    def _unique(items: Iterable[str]) -> Iterable[str]:
-        """Remove duplicates without changing order"""
-        seen = set()
-        output = []
-        for item in items:
-            if item not in seen:
-                output.append(item)
-                seen.add(item)
-        return output
-
-    # Remove empty and duplicate strings, and lowercase everything if need be
-    substrings = [s for s in substrings if s and len(s) > 0]
-    if not case_sensitive:
-        text = text.lower()
-        substrings = [s.lower() for s in substrings]
-    substrings = _unique(substrings)
-    offsets = []
-    for substring in substrings:
-        search_from = 0
-        # Search until one hit is found. Continue only if single_match is False.
-        while True:
-            start = text.find(substring, search_from)
-            if start == -1:
-                break
-            end = start + len(substring)
-            offsets.append((start, end))
-            if single_match:
-                break
-            search_from = end
+    nlp = spacy.blank(lang)
+    matcher = PhraseMatcher(nlp.vocab, attr="ORTH" if case_sensitive else "LOWER")
+    patterns = [nlp.make_doc(substring) for substring in substrings]
+    matcher.add("substrings", patterns)
+    # Perform the match
+    doc = nlp.make_doc(text)
+    matches: List[Span] = matcher(doc, as_spans=True)
+
+    def _filter_first_hit(matches: List[Span]) -> Iterable[Tuple[int, int]]:
+        seen_spans = []
+        filtered = []
+        for match in matches:
+            start, end = match.start_char, match.end_char
+            span_text = doc.char_span(start, end).text
+            if not case_sensitive:
+                span_text = span_text.lower()
+            if span_text not in seen_spans:
+                filtered.append((start, end))
+                seen_spans.append(span_text)
+        return filtered
+
+    offsets = (
+        _filter_first_hit(matches)
+        if single_match
```

Comment:
what's the point of single_match again?

Reply:
To be honest, I copied this over way back from the original [explosion/prodigy-openai-recipes](https://github.com/explosion/prodigy-openai-recipes) repo. My best guess: there may be tokens that have the same `ORTH` but function differently in a text (hence different entities) which I've also seen quite often in our internal project. However, I don't think `single_match` solves this problem one-to-one.